auraed/src/observe/observe_service.rs

```diff
@@ -70,13 +75,57 @@ impl ObserveService {
                 OsString::from("/sys/fs/cgroup"),
             ))),
             posix_signals,
-            sub_process_consumer_list: Vec::<Receiver<LogItem>>::new(),
+            sub_process_consumer_list: Arc::new(Mutex::new(HashMap::new())),
         }
     }
 
-    pub fn register_channel(&mut self, consumer: Receiver<LogItem>) {
-        info!("Added new channel");
-        self.sub_process_consumer_list.push(consumer);
+    pub async fn register_sub_process_channel(
+        &self,
+        pid: i32,
+        channel_type: LogChannelType,
+        channel: LogChannel,
+    ) -> Result<(), ObserveServiceError> {
+        info!("Registering channel for pid {pid} {channel_type:?}");
+        let mut consumer_list = self.sub_process_consumer_list.lock().await;
+        if consumer_list.get(&pid).is_none() {
+            let _ = consumer_list.insert(pid, HashMap::new());
+        }
+        if consumer_list
+            .get(&pid)
+            .expect("pid channels")
+            .get(&channel_type)
+            .is_some()
+        {
+            return Err(ObserveServiceError::ChannelAlreadyRegistered {
+                pid,
+                channel_type,
+            });
+        }
+        let _ = consumer_list
+            .get_mut(&pid)
+            .expect("pid channels")
+            .insert(channel_type, channel);
+        Ok(())
+    }
+
+    pub async fn unregister_sub_process_channel(
+        &self,
+        pid: i32,
+        channel_type: LogChannelType,
+    ) -> Result<(), ObserveServiceError> {
+        info!("Unregistering for pid {pid} {channel_type:?}");
+        let mut consumer_list = self.sub_process_consumer_list.lock().await;
+        if let Some(channels) = consumer_list.get_mut(&pid) {
+            if channels.remove(&channel_type).is_none() {
+                return Err(ObserveServiceError::ChannelNotRegistered {
+                    pid,
+                    channel_type,
+                });
+            }
+        } else {
+            return Err(ObserveServiceError::NoChannelsForPid { pid });
```

Comment:
Should we ignore these errors in the CellService? Does it make sense to return an error to the client when the executable has been stopped and the channels are not registered?

Reply:
perhaps not.. same as above.. maybe catch and warn instead?

Reply:
I think catch and warn is a good choice